### PR TITLE
Use Flush derives and generally update the Flush trait

### PR DESCRIPTION
**Pull request by Robbepop**
Update the `Flush` trait and use derive `Flush` proc. macros in `ink_core` and elsewhere.

Blocked by: https://github.com/paritytech/ink/pull/256

### CODE REVIEW

File: core/src/lib.rs [hunk 1]
```diff
@@ -58,3 +58,6 @@ pub mod env;
 pub mod env2;
 pub mod memory;
 pub mod storage;
+
+// Needed for derive macros of `core/derive` sub crate.
+pub(crate) use crate as ink_core;
```
Comment (ascjones) [inline core/src/lib.rs]: Very clever!

**Comment by Robbepop** [comment]
> LGTM. Why did we have all those empty impls of `fn flush` in the first place?

This is the question I have asked myself for a long time.

**Comment by ascjones** [comment]
CI failing

**Comment by Robbepop** [comment]
> CI failing

Fixed

**Pull request merged**